explainer-strings.md

```diff
@@ -0,0 +1,175 @@
+Note that it contains a `<style>` element. On the initial parse, the `<img>`
+element, which contains an event handler, is inside the `<style>` element and
+is thus being parsed as a (malformed) stylesheet. On the second parse, it
+will be parsed as a regular element, with an event handler.
+
+If this example is being presented to a sanitizer library, the sanitizer will
+see the result of the first pass, and will let the "event handler" - which it
+sees as style-sheet content - pass. If this is being re-parsed on inserting
+this into the DOM, the DOM parser will in turn see a regular element with an
+event handler.
+
+This kind of parse/re-parse mismatch (and its cousin, the parse context
+mismatch) has let to a number of attacks to bypass sanitizing libraryies,
+collectively known as mXSS.
+
+### Putting the Pieces Together: A New, Safer, and More Efficient API
+
+When contemplating these three concerns, we decided to
+reformulate the Sanitizer API along these rules:
+
+- The API will return the DOM tree it operates on, thus discouraging re-parsing
+  in general.
+- All string-based operations require a context node that will be used for
+  HTML parsing.
+- The API will preserve the context information, and won't hide it. Either
+  the API operates on the context node directly, or it will pass
+  along the context node in the result.
+
+This leads to the current API:
+
+For `Sanitizer.sanitize`, no strings and therefore no HTML parsing are
+involved. It does not require a parsing context. It returns a modified copy
+of the input.
+
+For `someElement.setHTML`, the context node is the `this` element that it is
+being called on. It returns no result at all, and instead the operation will
+commence on the `this` node.
+
+For `Sanitizer.sanitizeFor`, the developers will have to tell the Sanitizer
+what context they'd like to use the result in. The return value will contain
+both the result, and a newly created element that acts as the context node.
+Conveniently, a container for these two values already exists: It's the elment
+node`itself. Hence we return it, with the sanitized input as the children of the
+newly created context element.
+
+### An Escape Hatch
+
+What if a developer wishes to have a string-to-string API after all? Maybe
+because an existing code base was built around the assumption of a
+string-to-string sanitization library, and it isn'd easy to refactor?
+
+The `Sanitizer.sanitizeFor`method provides an escape hatch:
+`.sanitizeFor(value, context).innerHTML`.
+
+Given a suitable `context`, querying `.innerHTML` on the result of
+`Sanitizer.sanitizeFor` will retrieve the string result the developer
+presumably expects.  However, we recommend against this practice, as this
+opens up the app to context- and re-parse-mismatches and therefore mXSS
+attacks, and puts the burden to ensure safe usage entirely on
+the developer.
+
+# Please Try it Out
+
+Please try out the Sanitizer API. The [FAQ shows how to enable it in current
```

Comment:
Link to the playground at https://sanitizer-api.dev/?

Reply:
Done.